Sources/NIOHPACK/DynamicHeaderTable.swift

```diff
@@ -0,0 +1,169 @@
+    func dumpHeaders() -> String {
+        return self.storage.dumpHeaders()
+    }
+    
+    /// Searches the table for a matching header, optionally with a particular value. If
+    /// a match is found, returns the index of the item and an indication whether it contained
+    /// the matching value as well.
+    ///
+    /// Invariants: If `value` is `nil`, result `containsValue` is `false`.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the header for which to search.
+    ///   - value: Optional value for the header to find. Default is `nil`.
+    /// - Returns: A tuple containing the matching index and, if a value was specified as a
+    ///            parameter, an indication whether that value was also found. Returns `nil`
+    ///            if no matching header name could be located.
+    func findExistingHeader<C : Collection>(named name: C, value: C? = nil) -> (index: Int, containsValue: Bool)? where C.Element == UInt8 {
+        // looking for both name and value, but can settle for just name
+        // thus we'll search manually.
+        // TODO: there's probably a better algorithm for this.
+        var firstNameMatch: Int? = nil
+        for index in self.storage.indices(matching: name) {
+            if firstNameMatch == nil {
+                // record the first (most recent) index with a matching header name,
+                // in case there's no value match.
+                firstNameMatch = index
+            }
+            
+            if let value = value, self.storage.view(of: self.storage[index].value).matches(value) {
+                // this entry has both the name and the value we're seeking
+                return (index, true)
+            }
+        }
+        
+        // no value matches -- but did we find a name?
+        if let index = firstNameMatch {
+            return (index, false)
+        } else {
+            // no matches at all
+            return nil
+        }
+    }
+    
+    /// Appends a header to the table. Note that if this succeeds, the new item's index
+    /// is always zero.
+    ///
+    /// This call may result in an empty table, as per RFC 7541 § 4.4:
+    /// > "It is not an error to attempt to add an entry that is larger than the maximum size;
+    /// > an attempt to add an entry larger than the maximum size causes the table to be
+    /// > emptied of all existing entries and results in an empty table."
+    ///
+    /// - Parameters:
+    ///   - name: A collection of UTF-8 code points comprising the name of the header to insert.
+    ///   - value: A collection of UTF-8 code points comprising the value of the header to insert.
+    ///   - evictAutomatically: If `true`, the table will evict existing items to make room. Default is `true`.
+    /// - Returns: `true` if the header was added to the table, `false` if not.
+    mutating func addHeader<C : Collection>(named name: C, value: C, evictAutomatically: Bool = true) throws where C.Element == UInt8 {
```

Comment:
The actual intent was to be able to have the encoding side choose, rather than evicting, to handle the error directly by enlarging the header table. Judging by the spec, the table size is set initially using a SETTINGS frame, then can be adjusted upwards from that value using HEADERS frames, as in [RFC 7541 § 4.2](https://httpwg.org/specs/rfc7541.html#maximum.table.size). 

The idea is that the `HeaderIndexTable` and `HPACKEncoder` layers can (in the future, at least) make this visible to the clients via some simple API, and can use this flag to allow the error to bubble up to the point where it can insert a dynamic table size change at the start of the encoded header block in lieu of deleting things. There's plenty to think about regarding how to implement that and when to allow it, though, so for now the feature remains unused yet ready for implementation later.